charts/cryostat/templates/networkpolicy_ingress.yaml

```diff
@@ -0,0 +1,19 @@
+{{- if ((.Values.networkPolicy.ingress).enabled) }}
+apiVersion: networking.k8s.io/v1
+kind: NetworkPolicy
+metadata:
+  name: {{ .Release.Name }}-internal-ingress
+  namespace: {{ .Release.Namespace }}
+spec:
+  podSelector:
+    matchLabels:
+      {{- include "cryostat.selectorLabels" $ | nindent 6 }}
+  ingress:
+    - from:
+      - podSelector:
+          matchLabels:
+            {{- include "cryostat.selectorLabels" $ | nindent 12 }}
+        namespaceSelector:
+          matchLabels:
+            kubernetes.io/metadata.name: {{ .Release.Namespace }}
```

Comment:
> I tested with minikube with Calico as CNI and Nginx Ingress. I think the [default Kindnet](https://minikube.sigs.k8s.io/docs/handbook/network_policy/) is not actually implementing NetworkPolicies.

Yea, Kindnet doesn't implement them. That's supposed to mean they just get ignored and all traffic restrictions are lifted (or rather, are just not imposed)...

> I wonder if it would make sense to allow all ingress to Cryostat main pod instead?

At that point, the policies may as well just be disabled then. Maybe these should be turned off by default along with the egress policies, since they are just meant to be an additional level of protection.

Reply:
Would it make sense to use the NetworkPolicies to restrict traffic only for certain ports? Like for Cryostat talking to the storage pod? If the port is meant to be accessible from outside the cluster, then it's probably okay to be accessed from inside the cluster too.

Reply:
I'm not sure that I understand. Creating an Ingress NetworkPolicy basically blocks all incoming traffic, and then the various rules create an allowlist for traffic that is allowed through based on the origin of that traffic. Since there are no specifications on the ports in these rules currently, all ports are allowed so long as the origin Pod matches one of the rules. So I think it's already behaving like this: "If the port is meant to be accessible from outside the cluster, then it's probably okay to be accessed from inside the cluster too."

Reply:
I was thinking something like having one or more NetworkPolicy that:
* Always allows traffic to the OAuth Proxy container, no matter the source
* Filters traffic to the storage container/pod to only Cryostat itself

Reply:
There's a note in the OCP doc that @tthvo posted that mentions that NetworkPolicy objects are additive, so they can be used together to satisfy requirements like the above: https://docs.openshift.com/container-platform/4.16/networking/network_security/network_policy/about-network-policy.html#nw-networkpolicy-allow-from-router_about-network-policy

Reply:
Ah, I see. In that case I think that would mean the policies' selectors would only select the storage Pod, and not Cryostat's own Pod (which is what also contains the OAuth Proxy). That way all traffic to the Cryostat Pod is allowed, but traffic to storage would only be accepted if it originated from Cryostat.

Same thing would also be applied to the database.

Reply:
Right, they are additive, but the default when none are present is "allow everything". The moment the first one is defined for a Pod then the default policy for that Pod becomes "allow nothing", and the additive rules build the allowlist for what can come through. I think what you're describing would be a policy for the storage, a policy for the database, and no policy on Cryostat itself.

Reply:
That would work. Not sure if there are any ports in the Cryostat container we'd want to restrict. 

On another note, one use case I was thinking of for the operator would be for the new nginx container. We could have a NetworkPolicy to only allow traffic to it from pods in the target namespaces.

Reply:
I suppose we could still put a policy on the Cryostat Pod just for completeness, to make sure only the authproxy's HTTP(S) port(s) are externally visible.

Reply:
I guess splitting netpol would also be helpful for fine-grained egress rules in #209 :D